server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -1417,6 +1427,51 @@ public SearchPhase newSearchPhase(
         }
     }
 
+    private void extractCCSTelemetry(SearchResponse searchResponse, SearchTask searchTask) {
+        Map<String, CCSUsage.PerClusterUsage> clusterUsageMap = new HashMap<>();
+        for (String clusterAlias : searchResponse.getClusters().getClusterAliases()) {
+            SearchResponse.Cluster cluster = searchResponse.getClusters().getCluster(clusterAlias);
+            CCSUsage.PerClusterUsage clusterUsageInfo = new CCSUsage.PerClusterUsage(cluster.getTook());
+            clusterUsageMap.put(clusterAlias, clusterUsageInfo);
+        }
+
+        CCSUsage ccsUsage = new CCSUsage.Builder().took(searchResponse.getTookInMillis())
+            .async(isAsyncSearchTask(searchTask))
+            .minimizeRoundTrips(searchResponse.getClusters().isCcsMinimizeRoundtrips())
+            .numSkippedRemotes(searchResponse.getClusters().getClusterStateCount(SearchResponse.Cluster.Status.SKIPPED))
+            .perClusterUsage(clusterUsageMap)
+            .build();
+        usageService.getCcsUsageHolder().updateUsage(ccsUsage);
+    }
+
+    /**
+     * TransportSearchAction cannot access async-search code, so can't check whether this the Task
+     * is an instance of AsyncSearchTask, so this roundabout method is used
+     * @param searchTask SearchTask to analyze
+     * @return true if this is an async search task; false if a synchronous search task
+     */
+    private boolean isAsyncSearchTask(SearchTask searchTask) {
```

Comment:
does this need to be revised on the long run? If Async search is external, there should not ideally be references to it in ES core?

Reply:
Maybe we could have some facility maybe for the extending classes to inform about the fact they are using trackable feature (like async search). What is not clear is how to do it in practice - from what I understand, AsyncSearch just submits search task... Maybe add some method on `TransportSearchAction` since `TransportSubmitAsyncSearchAction` gets it injected? In general, the idea is to reverse the responsibility, so that `TransportSearchAction` would not need to be aware of extending classes, but I am not clear on how exactly it can be achieved yet.